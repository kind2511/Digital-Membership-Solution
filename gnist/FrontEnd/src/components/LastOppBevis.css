.last-opp-bevis-container {
  background-color: #ededed;
  border: 1px solid #ccc;
  border-radius: 8px;
  padding: 20px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  position: relative;  
}

.last-opp-bevis-container .section-title {
  color: #333;
  font-size: 20px;
  margin-bottom: 10px;
}

.last-opp-bevis-container .search-input {
  width: 95%;
  padding: 10px;
  margin-bottom: 10px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.section-content {
  background-color: #fff;
  width: 95%;
  padding: 10px;
  border-radius: 4px;
  max-height: 118px; 
  overflow-y: auto; 
} 

.search-status {
  color: #666;
  margin-bottom: 10px;
}

.section-content,
.selected-member {
  background-color: #fff;
  padding: 10px;
  border-radius: 4px;
}

.search-result,
.selected-name {
  padding: 10px;
  border-bottom: 1px solid #eee;
  cursor: pointer;
}

.search-result:hover,
.selected-name {
  background-color: #f7f7f7;
}

.search-result:last-child {
  border-bottom: none;
}

.buttons-container {
  display: flex;
  justify-content: flex-end;
  gap: 10px;
  margin-top: 10px;
}

.upload-button,
.close-button {
  padding: 8px 16px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.upload-button {
  background-color: #4CAF50;  
  color: white;
}

.close-button {
  background-color: #f44336;  
  color: white;
}

.last-opp-success-message {
  position: fixed; 
  left: 50%;
  bottom: 20px;
  transform: translateX(-50%);
  background-color: #28a745;
  color: white;
  text-align: center;
  padding: 10px;
  border-radius: 4px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
  z-index: 100;  
}

.last-opp-modal-container {
  display: flex;
  justify-content:center;
  align-items: center;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 1000;
}

.last-opp-modal-content {
  background-color: #fefefe;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  width: 50%;  
  min-height: 400px;  
}

.last-opp-close-modal {
  color: #aaa;
  float: right;
  font-size: 28px;
  font-weight: bold;
  cursor: pointer;
}

.last-opp-close-modal:hover,
.last-opp-close-modal:focus {
  color: black;
  text-decoration: none;
  cursor: pointer;
}

.vis-bevis-modal-button {
  color: #32325D;
}

.certificates-container {
  max-height: 350px;  
  overflow-y: auto;
}

.certificate-entry {
  padding: 10px;
  border-bottom: 1px solid #eee;
  cursor: pointer;
}

.certificate-entry:last-child {
  border-bottom: none;
}


.certificate-image {
  width: 100%;  
  max-height: 400px;  
  object-fit: contain;  
  margin-top: 20px;  
}


.last-opp-delete-button{
  padding: 9px;
  background-color:red;
}