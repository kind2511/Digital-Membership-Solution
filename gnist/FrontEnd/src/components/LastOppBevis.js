import React, { useState, useRef } from 'react';
import './LastOppBevis.css';

function LastOppBevis() {
    const [searchTerm, setSearchTerm] = useState('');
    const [selectedMember, setSelectedMember] = useState(null);
    const [results, setResults] = useState([]);
    const [searchStatus, setSearchStatus] = useState('');
    const [uploadSuccess, setUploadSuccess] = useState(false);
    const [deleteSuccess, setDeleteSuccess] = useState(false);
    const [showModal, setShowModal] = useState(false);
    const [certificates, setCertificates] = useState([]);
    const [imageViewer, setImageViewer] = useState('');
    const fileInputRef = useRef(null);
    const certificateNameRef = useRef(null);

    const fetchData = (value) => {
        if (value.trim() === '') {
            setResults([]);
            setSearchStatus('');
            return;
        }
        setSearchStatus('Searching...');
        fetch(`http://127.0.0.1:8000/digital_medlemsordning/search_member/?name=${value}`)
            .then(response => response.json())
            .then(data => {
                setResults(data);
                setSearchStatus(`Found ${data.length} results.`);
            })
            .catch(error => {
                console.error("Error fetching data:", error);
                setSearchStatus("Failed to fetch data. Please try again later.");
            });
    };

    const fetchCertificates = (auth0ID) => {
        fetch(`http://127.0.0.1:8000/digital_medlemsordning/get_member_certificates/${auth0ID}/`)
            .then(response => response.json())
            .then(data => {
                console.log('Fetched certificates:', data);
                setCertificates(data);
            })
            .catch(error => {
                console.error("Error fetching certificates:", error);
            });
    };

    const deleteCertificate = (certificateId) => {
        console.log('Deleting certificate with ID:', certificateId);
        if (certificateId === undefined) {
            console.error('Certificate ID is undefined.');
            return;
        }
        fetch(`http://127.0.0.1:8000/digital_medlemsordning/delete_member_certificate/${certificateId}/`, {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            }
        })
        .then(response => {
            console.log('Response status:', response.status);
            if (!response.ok) {
                throw new Error(`Network response was not ok: ${response.statusText}`);
            }
            return response.json().catch(() => {});
        })
        .then(data => {
            console.log('Success data:', data);
            setDeleteSuccess(true);
            setTimeout(() => setDeleteSuccess(false), 3000);
            fetchCertificates(selectedMember.auth0ID);
        })
        .catch(error => {
            console.error("Error deleting certificate:", error);
            alert('Failed to delete certificate.');
        });
    };

    const handleSearchChange = (e) => {
        const value = e.target.value;
        setSearchTerm(value);
        fetchData(value);
    };

    const handleSelectMember = (member) => {
        setSelectedMember(member);
        setSearchTerm(`${member.first_name} ${member.last_name}`);
        setResults([]);
        setSearchStatus('');
        fetchCertificates(member.auth0ID);
    };

    const handleUpload = () => {
        if (!selectedMember) {
            alert('Velg et medlem å laste opp et bevis for.');
            return;
        }
        const file = fileInputRef.current.files[0];
        const certificateName = certificateNameRef.current.value;
        if (!file || !certificateName) {
            alert('Velg en fil og skriv inn et bevis navn for å laste opp.');
            return;
        }
        const formData = new FormData();
        formData.append('certificate_image', file);
        formData.append('certificate_name', certificateName);

        fetch(`http://127.0.0.1:8000/digital_medlemsordning/upload_member_certificates/${selectedMember.auth0ID}/`, {
            method: 'POST',
            body: formData,
            headers: {
                'Accept': 'application/json',
            },
        })
        .then(response => {
            if (!response.ok) {
                throw new Error('Network response was not ok.');
            }
            return response.json();
        })
        .then(data => {
            console.log('Success:', data);
            setUploadSuccess(true);
            setTimeout(() => setUploadSuccess(false), 3000);
            fetchCertificates(selectedMember.auth0ID); 
        })
        .catch(error => {
            alert('Failed to upload certificate.');
            console.error('Error:', error);
        });
    };

    const handleClose = () => {
        setSelectedMember(null);
        setSearchTerm('');
        setSearchStatus('');
        setCertificates([]); 
        setImageViewer(''); 
    };

    const toggleModal = () => {
        setShowModal(!showModal);
        setImageViewer(''); 
    };

    const handleCertificateClick = (cert) => {
        setImageViewer(cert.certificate_image); 
    };

    return (
        <div className="last-opp-bevis-container">
            <h2 className="section-title">Last Opp Bevis</h2>
            {!selectedMember && (
                <input
                    type="text"
                    className="search-input"
                    placeholder="Søk etter medlem..."
                    value={searchTerm}
                    onChange={handleSearchChange}
                />
            )}
            <div className="search-status">{searchStatus}</div>
            {selectedMember ? (
                <div className="selected-member">
                    <div className="selected-name">{`${selectedMember.first_name} ${selectedMember.last_name}`}</div>
                    <input type="file" ref={fileInputRef} className="file-input" />
                    <input type="text" ref={certificateNameRef} placeholder="Skriv inn bevis navn" className="certificate-name-input" />
                    <div className="buttons-container">
                        <button className="upload-button" onClick={handleUpload}>Last Opp</button>
                        <button className="close-button" onClick={handleClose}>Lukk</button>
                        <button className="vis-bevis-modal-button" onClick={toggleModal}>Vis Andre Bevis</button>
                    </div>
                </div>
            ) : (
                <div className="section-content">
                    {results.map((member, index) => (
                        <div key={index} className="search-result" onClick={() => handleSelectMember(member)}>
                            {`${member.first_name} ${member.last_name}`}
                        </div>
                    ))}
                </div>
            )}
            {uploadSuccess && (
                <div className="last-opp-success-message">Bevis ble lastet opp</div>
            )}
            {deleteSuccess && (
                <div className="last-opp-success-message">Bevis ble slettet</div>
            )}
            {showModal && (
                <div className="last-opp-modal-container">
                    <div className="last-opp-modal-content">
                        <span className="last-opp-close-modal" onClick={toggleModal}>&times;</span>
                        <div className="certificates-container">
                            {certificates.map((cert, index) => (
                                <div key={cert.certificateID} className="certificate-entry">
                                    <div onClick={() => handleCertificateClick(cert)}>
                                        {`${index + 1}. ${cert.certificate_name}`}
                                    </div>
                                    <button className="last-opp-delete-button" onClick={() => {
                                        console.log('Certificate ID to be deleted:', cert.certificateID);
                                        deleteCertificate(cert.certificateID);
                                    }}>Slett</button>
                                </div>
                            ))}
                        </div>
                        {imageViewer && <img src={imageViewer} alt="Certificate View" className="certificate-image" />}
                    </div>
                </div>
            )}
        </div>
    );
}

export default LastOppBevis;
