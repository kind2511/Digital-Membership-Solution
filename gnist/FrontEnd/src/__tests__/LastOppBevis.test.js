import React from 'react';
import renderer from 'react-test-renderer';
import LastOppBevis from '../components/LastOppBevis';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

// Mock axios for the test
const mock = new MockAdapter(axios);
const membersMock = [
  {
    auth0ID: 'auth0|123456789',
    first_name: 'John',
    last_name: 'Doe'
  },
  {
    auth0ID: 'auth0|987654321',
    first_name: 'Jane',
    last_name: 'Doe'
  }
];
const certificatesMock = [
  {
    certificateID: 1,
    certificate_name: 'Certificate 1',
    certificate_image: '/path/to/certificate1.jpg'
  },
  {
    certificateID: 2,
    certificate_name: 'Certificate 2',
    certificate_image: '/path/to/certificate2.jpg'
  }
];

mock.onGet('http://127.0.0.1:8000/digital_medlemsordning/search_member/?name=').reply(200, membersMock);
mock.onGet('http://127.0.0.1:8000/digital_medlemsordning/get_member_certificates/auth0|123456789/').reply(200, certificatesMock);
mock.onDelete('http://127.0.0.1:8000/digital_medlemsordning/delete_member_certificate/1/').reply(200);
mock.onDelete('http://127.0.0.1:8000/digital_medlemsordning/delete_member_certificate/2/').reply(200);

test('LastOppBevis component snapshot', () => {
  const component = renderer.create(<LastOppBevis />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
