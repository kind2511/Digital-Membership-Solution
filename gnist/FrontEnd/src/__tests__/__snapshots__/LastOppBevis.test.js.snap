// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastOppBevis component snapshot 1`] = `
<div
  className="last-opp-bevis-container"
>
  <h2
    className="section-title"
  >
    Last Opp Bevis
  </h2>
  <input
    className="search-input"
    onChange={[Function]}
    placeholder="SÃ¸k etter medlem..."
    type="text"
    value=""
  />
  <div
    className="search-status"
  />
  <div
    className="section-content"
  />
</div>
`;
